### PR TITLE
Fixes color pipeline to match exrdisplay math.

### PR DESCRIPTION
**Pull request by ggarra13**
The gamma function needs to be applied as part of the exposure math.  Note, however, that it makes the levels' checks useless.  To fix them, they also would need to move to the exposureFunc() function.

### CODE REVIEW

**Comment by ggarra13** [comment]
Another issue with the way exrdisplay does its math is that it relies on a gamma of, usually, 2.2.  This conflicts with the OCIO pipeline when using an sRGB view/display, as you'd get a (close) to 2.2 gamma applied twice.

**Comment by darbyjohnston** [comment]
Would you want to mix the exrdisplay settings and OCIO, it seems like they would be used separately?

**Comment by ggarra13** [comment]
On 6/10/22 19:53, Darby Johnston wrote:
>
> Would you want to mix the exrdisplay settings and OCIO, it seems like 
> they would be used separately?
>
If one of the parameters is called exposure, the user will certainly 
expect it to use it with or without OCIO to lower or raise the 
'brightness' by 1 f-stop increments.  If you have a toggle to turn it on 
and off as you do now, the user (or at least me), will expect the 
default values to be a no-op.

What I expect the user to not use when using OCIO is the gamma setting, 
as that is now, more accurately, baked in the display/view transform.

The issue with the exrdisplay pipeline is that, albeit more physically 
correct, it predates OCIO, so now I am not sure how you integrate it 
properly.

-- 
Gonzalo Garramuño
***@***.***

**Comment by ggarra13** [comment]
I will leave it to your judgement here on how you want to deal with it.  With mrViewer, I won't expose the exposure parameters as I think that leads to confusion. I'll stick with gain and apply it as a multiplier to brightness as you suggested.  The GUI currently shows the current fstop and allow adjusting the gain in 1/2 fstop increments, like nuke does.

**Comment by darbyjohnston** [comment]
Hi, and sorry for the delay! Maybe it would be good to rename the ```Exposure``` struct to something like ```LegacyEXRDisplay``` so that it is more clear what it does. I would like to keep it for comparison to exrdisplay but it is probably more appropriate for debugging uses rather than as user controls.

**Comment by ggarra13** [comment]
Sounds fine by me.

File: lib/tlGL/RenderShaders.cpp [hunk 1]
```diff
@@ -333,12 +333,12 @@ namespace tl
                 "    tmp[0] = (value[0] - data.inLow) / data.inHigh;\n"
                 "    tmp[1] = (value[1] - data.inLow) / data.inHigh;\n"
                 "    tmp[2] = (value[2] - data.inLow) / data.inHigh;\n"
-                "    if (tmp[0] >= 0.0)\n"
-                "        tmp[0] = pow(tmp[0], data.gamma);\n"
-                "    if (tmp[1] >= 0.0)\n"
-                "        tmp[1] = pow(tmp[1], data.gamma);\n"
-                "    if (tmp[2] >= 0.0)\n"
-                "        tmp[2] = pow(tmp[2], data.gamma);\n"
+                // "    if (tmp[0] >= 0.0)\n"
```
Comment (darbyjohnston) [inline lib/tlGL/RenderShaders.cpp:336]: The exposure controls are meant to be used as an alternative to the other controls, so this code should be left in place. After this PR is finished I will rename the exposure code to something like "exrdisplay" and add a comment about it's usage.

File: lib/tlGL/RenderVideo.cpp [hunk 2]
```diff
@@ -519,7 +519,7 @@ namespace tl
                 p.shaders["display"]->setUniform("levelsEnabled", displayOptions.levelsEnabled);
                 p.shaders["display"]->setUniform("levels.inLow", displayOptions.levels.inLow);
                 p.shaders["display"]->setUniform("levels.inHigh", displayOptions.levels.inHigh);
-                p.shaders["display"]->setUniform("levels.gamma", displayOptions.levels.gamma > 0.F ? (1.F / displayOptions.levels.gamma) : 1000000.F);
+                p.shaders["display"]->setUniform("levels.gamma", 1.0F );
```
Comment (darbyjohnston) [inline lib/tlGL/RenderVideo.cpp:522]: This code should be left as well.

File: lib/tlGL/RenderShaders.cpp [hunk 3]
```diff
@@ -373,9 +373,12 @@ namespace tl
                 "        value[1] = data.k + knee(value[1] - data.k, data.f);\n"
                 "    if (value[2] > data.k)\n"
                 "        value[2] = data.k + knee(value[2] - data.k, data.f);\n"
-                "    value[0] *= 0.332;\n"
-                "    value[1] *= 0.332;\n"
-                "    value[2] *= 0.332;\n"
+                "    if (value[0] > 0.0 ) value[0] = pow(value[0], data.g );\n"
+                "    if (value[1] > 0.0 ) value[1] = pow(value[1], data.g );\n"
+                "    if (value[2] > 0.0 ) value[2] = pow(value[2], data.g );\n"
+                "    value[0] *= pow(2, -3.5 * data.g );\n"
```
Comment (darbyjohnston) [inline lib/tlGL/RenderShaders.cpp:379]: Are these numbers from the exrdisplay documentation?

**Comment by darbyjohnston** [comment]
OK, thanks; if you can make the other changes I'll go ahead and merge this.

**Comment by ggarra13** [comment]
I'll open a new clean PR as I modified the branch twice.

**Pull request closed without merging**